src/back/msl/writer.rs

```diff
@@ -1306,14 +1307,38 @@ impl<W: Write> Writer<W> {
             match *statement {
                 crate::Statement::Emit(ref range) => {
                     for handle in range.clone() {
+                        if let Some(name) =
+                            context.expression.function.named_expressions.get(&handle)
+                        {
+                            write!(self.out, "{}", level)?;
+
+                            // Front end provides names for all variables at the start of writing.
+                            // But we write them to step by step. We need to recache them inside `write_named_expr`
+                            // Otherwise, we could accidentally write variable name instead of full expression.
+                            // Also, we use sanitized names! It defense backend from generating variable with name from reserved keywords.
+                            let sanitized_name = self.namer.call_unique(name);
+                            self.start_baking_expression(
+                                handle,
+                                &context.expression,
+                                &sanitized_name,
+                            )?;
+                            self.put_expression(handle, &context.expression, true)?;
+                            self.named_expressions.insert(handle, sanitized_name);
+
+                            writeln!(self.out, ";")?;
+                            continue;
```

Comment:
Can we avoid control flow here, please?
Another thing, can we merge this with the other bit of code here?
I.e. something like
```
let name = match context.expression.function.named_expressions.get(&handle) {
  Some(name) => Some(self.namer.call_unique(name)),
  None => {
    let min_ref_count = ...;
    if min_ref_count <= ... {
      Some(format!(...))
    } else {
      None
    }
  };
  if let Some(name) = name {
    // spew out the expression
  }
}
```

Reply:
Great idea, thank you! Fixed.